### DISCUSSION TITLE
installing pip packages on airflow image/helm chart?

### DISCUSSION THREAD

**Discussion by tsaaii** [apache/airflow#20689] [Q&A] [answered]
I am looking to add snowflake operator to airflow [following this tutorial here.](https://community.snowflake.com/s/article/How-to-connect-Apache-Airflow-to-Snowflake-and-schedule-queries-jobs#:~:text=Configure%20snowflake%20connection%20in%20Airflow&text=One%20can%20run%20below%20commands%20after%20activating%20the%20python%20virtual%20enviroment.&text=3.,details%20as%20shown%20in%20screenshot.)

I have a docker file and I am using a helm chart to install airflow.

For this, I have to add few python packages and I tried these 3 ways of doing it.

1. adding it to dockerfile and building the image and then use helm chart on it to install airflow.

```
FROM apache/airflow:latest
USER root
RUN pip install "package1" "package2" "package3" 
```

2. adding them to helm chart values.yaml file under extraPipPackages

```
executor: "KubernetesExecutor"
images:
  airflow:
    repository: "local/airflow"
    tag: "latest"
    extraPipPackages:
      - "Package1==1.3" 
      - "package2==2.0" 
```

3. adding a command under webserver in values.yaml file 
```
webserver:
   command: ["pip install package1 && pip install package2"]
```

I tried all the 3, while no method gives an error, i don't see my pip packages( this is supposed to show snowflake operator in airflow)

docker logs while building image shows that these packages are being installed but I don't see them when I look in the container

`docker exec container_id pip list` doesn't show my packages which were installed earlier.

What is the right way of installing pip packages?

My kubectl version is

$ kubectl version
```
Client Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.0", GitCommit:"ab69524f795c42094a6630298ff53f3c3ebab7f4", GitTreeState:"clean", BuildDate:"2021-12-07T18:16:20Z", GoVersion:"go1.17.3", Compiler:"gc", Platform:"windows/amd64"}
Server Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.4", GitCommit:"b695d79d4f967c403a96986f1750a35eb75e75f1", GitTreeState:"clean", BuildDate:"2021-11-17T15:42:41Z", GoVersion:"go1.16.10", Compiler:"gc", Platform:"linux/amd64"}
```
My helm chart version is 

`version.BuildInfo{Version:"v3.7.1", GitCommit:"1d11fcb5d3f3bf00dbe6fe31b8412839a96b3dc4", GitTreeState:"clean", GoVersion:"go1.16.9"}`

What is the right way to add pip packages?

**Comment by potiuk** [comment, marked as answer]
The best it method 1) (but properly -  you should use airflow user not root). Never, ever use 2 nor 3 (actually [Airflow Official  Helm Chart](https://airflow.apache.org/docs/helm-chart/stable/index.html) does not allow to use theml)  - those are extremely bad antipattern and you might find some issues and discussions why - see also [this warning](https://airflow.apache.org/docs/docker-stack/entrypoint.html#installing-additional-requirements).

You can find all the examples and explanation on how you can extend or customize Airlfow  image [here](https://airflow.apache.org/docs/docker-stack/build.html).

**Reply by potiuk** [reply]
BTW: in the official chart you even have it explained in "production guide": https://airflow.apache.org/docs/helm-chart/stable/production-guide.html#extending-and-customizing-airflow-image

**Reply by tsaaii** [reply]
Thank you. I could get it working.

**Reply by subhosuper** [reply]
Hello @tsaaii  @potiuk  
I am using the official helm chart of airflow, Airflow version 2.3.0.
I have build my custom image where I have installed some python packages.
And I am hosting it on a private docker registry.
But the problem is in the official helm chart there is no way which **I could find where to put the image pull secrets** where I can mention the credentials of the private docker registry.
Can you please help me with this?
I am using this values.yaml [https://github.com/apache/airflow/blob/main/chart/values.yaml](url)

**Reply by potiuk** [reply]
https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/

**Comment by boukisge** [comment]
I am also planning to install pip packages using dockerfile ( approach 1) . The issue that I am having is that pip list returns nothing , no error message.

I have posted details in slack channel - Apache Airflow Community  - troubleshooting

I have deployed Airflow official image v2.1.2 using helm chart community edition and is up and running no issues. I am trying to get the list of installed packages from inside the container using pip list in preparation of adding additional packages . Pip list command outputs nothing , no error message.  Pip show about a specific package works fine ( ex. pip show wheel) . When I list the contents of  directory (/usr/local/lib/python3.7/site-packages)  using ls command, I do see 2 packages installed (wheel, pip) . When I run the command   pip --version returns ( pip 21.2.4 from /usr/local/lib/python3.7/site-packages/pip (python 3.7) .  I have deployed other apps in the same k8s cluster based on a python base image and pip list works fine inside the container

**Reply by potiuk** [reply]
You need to use the entrypoint of aiflow to see the `pip list`. Airflow image has alll packages installed in the `~/.local` directory. All packages are installed  with `pip --user` flag. This flag is enabled by default `PIP_USER="true"` and HOME directory properly set when you enter airflow via its entrpoint `/entrypoint`. This happens by default when you just run `docker run` and pass bash as parameter (see https://airflow.apache.org/docs/docker-stack/entrypoint.html#executing-commands for details).

```
[jarek:~/code/airflow-clone] v2-2-test 3s 130 ± docker run -it apache/airflow:2.2.3 bash
                                                                            
airflow@c1d5c8317459:/opt/airflow$ pip list   
Package                                  Version
---------------------------------------- ----------
adal                                     1.2.7  
alembic                                  1.7.5 
amqp                                     5.0.7 
anyio                                    3.4.0
apache-airflow                           2.2.3
apache-airflow-providers-amazon          2.4.0 
apache-airflow-providers-celery          2.1.0 
apache-airflow-providers-cncf-kubernetes 2.2.0
```

This is all described in https://airflow.apache.org/docs/docker-stack/entrypoint.html also explaining why this is needed (OpenShift Compatibility and following the OpenShift Guidelines), The entrypoint is `/entrypoint`. 

You are likely using `/bin/bash` as the entrypoint and there you get only "system dependencies".

BTW. It is considered a bad practice to install dependencies in system provided site packages (/usr/local/lib/python3.7/site-packages/
). You should never do that. While we differ with recommendation on that with PIP maintainers (huge discussion with a lot oh heated opinions in https://github.com/pypa/pip/issues/10556), I think I agree with PIP maintainers that installing deps to system site packages is a bad idea. This is one of the reasons --user flag is used in Airflow image. 

There is an upcoming PEP668 https://www.python.org/dev/peps/pep-0668/ which is all about it but it will likely take years until it will get approved and adopted, so until then either using virtualenv (as recommented by PIP  maintainers) or using `--user` flag (I believe it is much better for docker containers case) is the right solution.

**Reply by potiuk** [reply]
Also if you want to **just** extend the image, follow the examples in https://airflow.apache.org/docs/docker-stack/build.html#extending-the-image

**Reply by boukisge** [reply]
I am also getting a WARNING , not sure if its related
WARNING: The directory '/.cache/pip' or its parent directory is not owned or is not writable by the current user. The cache has been disabled

**Reply by boukisge** [reply]
I am able to install additional packages using dockerfile when i build the image , i just want to verify the list

**Reply by potiuk** [reply]
> I am able to install additional packages using dockerfile when i build the image , i just want to verify the list

See above where I explained it with examples on how you can do it and why it works this way when you are, wrongly, not using the default entrypoint.

**Reply by boukisge** [reply]
Thank You . I will give it a try

**Reply by boukisge** [reply]
Running docker run command is forbidden in our  organization, instead the OpenShift debug command is used

in this case

oc debug airflow-web-9b44445b
"Starting pod/airflow-web-9b44445b-debug, command was: /usr/bin/dumb-init -- /entrypoint bash -c exec airflow webserver"
Apologies for my lack of understanding . 
What are the steps/commands required to display installed packages

**Reply by potiuk** [reply]
Once you are in the shell in the pod, just running `/entrypoint bash` should do what you want (and set the right environment).

**Reply by potiuk** [reply]
Explanation: I do not know details of `oc debug` but from https://www.redhat.com/sysadmin/how-oc-debug-works it looks like it will drop you in the container and bypass the entrypoint (and give you root access no matter what). But by running `/entrypoint bash` you will manually set the same environment that your application gets - as you see the application runs the entrypoint:

```
/usr/bin/dumb-init -- /entrypoint bash -c exec airflow webserver
```

**Reply by boukisge** [reply]
it worked . thank you

**Reply by potiuk** [reply]
BTW. If you do make it works, I would really appreciate if you make a PR in our documentation:

I guess this is the best one: https://airflow.apache.org/docs/helm-chart/stable/production-guide.html  - adding a chapter about debugging on OpenShift. Airflow is created by > 1800 contributors and adding such documenation might be the first step to become one (and give back for the free software you get). It's super easy - just click "Sugest a change on this page" bottom-right button and you will be able to create Pull Request directly from there.

**Reply by potiuk** [reply]
Correction. Just checked - more than 1900 contributors already :)

**Reply by kobethuwis** [reply]
@tsaaii Here's a piece of my dockerfile, might help
```
WORKDIR ${AIRFLOW_HOME}
USER ${AIRFLOW_UID}
RUN         python${PYTHON_VERSION%.*} -m pip install --no-cache-dir pip botocore boto3 polling simplejson\
            && python${PYTHON_VERSION%.*} -m pip install --no-cache-dir apache-airflow-providers-apache-livy==${PROVIDERS_LIVY_VERSION}

```

**Comment by tsaaii** [comment]
@boukisge Yes this means, your pip packages were not installed on the image. Can you try 

`docker system prune --all`
`helm uninstall airflow -n airflow`

and retry this again?

**Comment by boukisge** [comment]
I have uninstall several times.

**Comment by boukisge** [comment]
Pip itself and wheel - at minimum should be displayed by running the pip
list command.

When I try pip show wheel - it works fine

**Comment by boukisge** [comment]
Yes , will do , what subject would you like to see?

**Reply by potiuk** [reply]
Choose one that your past-self would find easiest to find when looking for it :) . Really - you are the best person in the world to propose it as you struggled with it.

**Comment by boukisge** [comment]
Cool, thanks!

**Reply by subhosuper** [reply]
Hello @tsaaii @potiuk
I am using the official helm chart of airflow, Airflow version 2.3.0.
I have build my custom image where I have installed some python packages.
And I am hosting it on a private docker registry.
But the problem is in the official helm chart there is no way which **I could find where to put the image pull secrets where I can mention the credentials of the private docker registry** to pull the newly built Airflow image (where the python packages are present)
Can you please help me with this?
I am using this values.yaml [https://github.com/apache/airflow/blob/main/chart/values.yaml](https://github.com/apache/airflow/discussions/url)

**Reply by clementchane** [reply]
Hello @subhosuper 
I have the same concern right now. Could you please let me know how you managed to install airflow helm chart with a private docker repository image?
I do not know how to add the "imagePullSecrets" parameter.
That would be super helpful!
Thanks!

**Reply by potiuk** [reply]
I do not know if there is a way - probably you can take a look at `extraSecrets` for example (but I am not 100% sure). You have to take a look .
 
However, Helm chart is not only something you can use but - similarly to the image - something you can extend if you miss something. The easiest way to extend the Helm Chart if you want to add new resources which are not there is pretty similar to what you do with image. Make your own. This is what Airflow image does with Postgres chart when internal Postgres is enabled.

1) Create your own Helm Chart with resources you need.
2) Make Airflow chart a subchart of yours

That's it.

**Reply by subhosuper** [reply]
> Hello @subhosuper I have the same concern right now. Could you please let me know how you managed to install airflow helm chart with a private docker repository image? I do not know how to add the "imagePullSecrets" parameter. That would be super helpful! Thanks!

Hey @clementchane  there is a way, the `registry` directive is there in the values.yaml.
Search for this in the values.yaml: 

"Auth secret for a private registry
This is used if pulling airflow images from a private registry

registry:
  secretName:"

Mention your private registry secret name as value of "secretName"

And search for this:

"images:
  airflow:
    repository: repository.docker.com/airflow
    tag: tag23
    pullPolicy: IfNotPresent"

Mention your repo, tag and pull policy here.

**Reply by clementchane** [reply]
Oh yes perfect thank you! It worked
